### PR TITLE
Add X-Forwarded-SSL and X-Forwarded-Port to appaccess.

### PR DESCRIPTION
**Pull request by mdwn**
Application Access now adds in X-Forwarded-Ssl and X-Forwarded-Port headers. Tests have been added and adjusted to look for these new headers as well.

Notes: So I uncovered a few things while adding these new headers to websockets:

1. The oxy headers were *not* being pushed into websocket requests before, as the original writer was being overwritten.
2. The current header rewriter in oxy (as of this writing) will not work with websockets, as it will remove the websocket upgrade headers.
3. This PR depends on https://github.com/gravitational/oxy/pull/7.

### CODE REVIEW

File: lib/srv/app/header_rewriter.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	port := "80"
+	if req.Header.Get(forward.XForwardedProto) == "https" {
+		port = "443"
+	}
+
+	if req.URL.Port() != "" {
+		req.Header.Set(common.XForwardedPort, req.URL.Port())
+	} else {
+		req.Header.Set(common.XForwardedPort, port)
+	}
```
Comment (ryanclark) [inline lib/srv/app/header_rewriter.go:57]:
```suggestion
	if req.URL.Port() != "" {
		req.Header.Set(common.XForwardedPort, req.URL.Port())
	} else {
		// Guess some default ports as the port is not explicitly set in the request.
		port := "80"
		if req.Header.Get(forward.XForwardedProto) == "https" {
			port = "443"
		}
		
		req.Header.Set(common.XForwardedPort, port)
	}
```

I meant something more like this, as `port` isn't needed unless `req.URL.Port()` is `""`.

Comment (mdwn) [reply lib/srv/app/header_rewriter.go:57]: Ahhhh, gotcha. Sorry, brain fart there.

**Review by ryanclark — APPROVED** [review]

**Comment by ryanclark** [comment]
What's your thoughts on updating our fork of `oxy` to include this instead?

LGTM anyhow

**Comment by mdwn** [comment]
> What's your thoughts on updating our fork of `oxy` to include this instead?
> 
> LGTM anyhow

I had a PR for that :-) https://github.com/gravitational/oxy/pull/4. There's some controversy over oxy at the moment due to its semi-maintained nature as part of Teleport, and it's different enough that we can't just totally depend on upstream (maybe).

I think that, for better or for worse, that's a weightier discussion than this task/PR allows for.

File: lib/srv/app/session.go [hunk 2]
```diff
@@ -171,12 +172,25 @@ func (s *Server) withJWTTokenForwarder(ctx context.Context, sess *sessionChunk,
 		return trace.Wrap(err)
 	}
 
+	// Duplicate the default behavior for the header rewriter in oxy, as there
+	// is no getters/setters for this in the Forwarder so that we can grab it
+	// and re-set it there.
+	h, err := os.Hostname()
+	if err != nil {
+		h = "localhost"
+	}
```
Comment (r0mant) [inline lib/srv/app/session.go:181]: This looks out of place here. Can we just expose a `NewHeaderRewriter` constructor in oxy that would encapsulate this logic?

File: lib/srv/app/session.go [hunk 3]
```diff
@@ -171,12 +172,25 @@ func (s *Server) withJWTTokenForwarder(ctx context.Context, sess *sessionChunk,
...
+	hr := &headerRewriter{
+		delegate: delegate,
+	}
 	sess.fwd, err = forward.New(
 		forward.FlushInterval(100*time.Millisecond),
 		forward.RoundTripper(transport),
 		forward.Logger(logrus.StandardLogger()),
 		forward.WebsocketRewriter(transport.ws),
 		forward.WebsocketDial(transport.ws.dialer),
+		forward.Rewriter(hr),
```
Comment (r0mant) [inline lib/srv/app/session.go:193]: Couple questions here:

- Is this the only place we need to set our custom rewriter? We also create forwarder in web/apps.
- What about websocket requests rewriter?

File: integration/appaccess/fixtures.go [hunk 4]
```diff
@@ -338,6 +338,8 @@ var forwardedHeaderNames = []string{
 	"X-Forwarded-Host",
 	"X-Forwarded-Server",
 	"X-Forwarded-For",
+	"X-Forwarded-SSL",
```
[the remark this answers is not in the document]
Comment (mdwn) [reply integration/appaccess/fixtures.go:341]: Ah dang, yeah it's supposed to be `Ssl`.

File: lib/srv/app/session.go (hunk 2 above)
Comment (mdwn) [reply lib/srv/app/session.go:181]: Yes we can. I'll submit that PR shortly.

File: lib/srv/app/session.go (hunk 3 above)
Comment (mdwn) [reply lib/srv/app/session.go:193]: Let me take a look at those. Thanks for letting me know about those.

Comment (mdwn) [reply lib/srv/app/session.go]: I've addressed this, but there's some more explanation in the PR description.

**Comment by mdwn** [comment]
@r0mant @codingllama Ping on reviews for this one.

**Review by r0mant — APPROVED** [review]

File: lib/srv/app/common/header.go [hunk 5]
```diff
@@ -24,14 +24,22 @@ import (
...
-	forward.XForwardedHost,
-	forward.XForwardedProto,
-	forward.XForwardedServer,
+	XForwardedSSL,
+}
+
+func init() {
+	// Append the rest of oxy's headers into the reserved headers.
+	ReservedHeaders = append(ReservedHeaders, forward.XHeaders...)
 }
```
[the remark this answers is not in the document]
Comment (mdwn) [reply lib/srv/app/common/header.go]: You're right, it doesn't!

**Pull request merged**